common/setups/rasr/rasr_system.py

```diff
@@ -0,0 +1,225 @@
+    ):
+        self.crp[corpus_key].language_model_config = rasr.RasrConfig()
+        self.crp[corpus_key].language_model_config.type = type
+        self.crp[corpus_key].language_model_config.file = filename
+        self.crp[corpus_key].language_model_config.scale = scale
+
+    @tk.block()
+    def _init_lexicon(
+        self, corpus_key: str, filename: Path, normalize_pronunciation: bool, **kwargs
+    ):
+        self.crp[corpus_key].lexicon_config = rasr.RasrConfig()
+        self.crp[corpus_key].lexicon_config.file = filename
+        self.crp[
+            corpus_key
+        ].lexicon_config.normalize_pronunciation = normalize_pronunciation
+
+    @staticmethod
+    def _assert_corpus_name_unique(*args):
+        name_list = []
+        for i in args:
+            name_list.extend(list(i.keys()))
+        assert len(name_list) == len(set(name_list)), "corpus names are not unique"
+
+    def add_corpus(self, corpus_key: str, data: RasrDataInput, add_lm: bool):
+        self.corpora[corpus_key] = data.corpus_object
+        self.concurrent[corpus_key] = data.concurrent
+        self._init_corpus(corpus_key)
+        self._init_lexicon(corpus_key, **data.lexicon)
+        if add_lm:
+            self._init_lm(corpus_key, **data.lm)
+
+    @tk.block()
+    def extract_features(self, feat_args: dict, **kwargs):
+        corpus_list = self.train_corpora + self.dev_corpora + self.test_corpora
+
+        for k, v in feat_args.items():
+            for c in corpus_list:
+                if k == "mfcc":
+                    self.mfcc_features(c, **v)
+                if k == "gt":
+                    self.gt_features(c, **v)
+                if k == "fb":
+                    self.fb_features(c, **v)
+                if k == "energy":
+                    self.energy_features(c, **v)
+                if k == "voiced":
+                    self.voiced_features(c, **v)
+                if k == "plp":
+                    self.plp_features(c, **v)
+                if k == "tone":
+                    self.tone_features(c, **v)
+                if k not in ("mfcc", "gt", "fb", "energy", "voiced", "tone", "plp"):
+                    self.generic_features(c, k, **v)
+            if k == "mfcc":
+                for trn_c in self.train_corpora:
+                    self.normalize(trn_c, "mfcc+deriv", corpus_list)
+
+        for kk in feat_args.keys():
+            if "energy" in feat_args.keys():
+                for t in self.train_corpora:
+                    if kk == "mfcc":
+                        self.add_energy_to_features(t, "mfcc+deriv")
+                        self.add_energy_to_features(t, "mfcc+deriv+norm")
+                    if kk == "gt":
+                        self.add_energy_to_features(t, "gt")
+                    if kk == "fb":
+                        self.add_energy_to_features(t, "fb")
+
+    # -------------------- Single Density Mixtures --------------------
+
+    def single_density_mixtures(
```

Comment:
This is used to estimate the prior from the alignment in Hybrid DNN-HMM systems when _the Povey method_ is not used. So Yes.

Reply:
> Or is this needed ONLY for DNN-HMM and can/should be moved to the GMM system?

This is needed for all _Hybrid_ systems, where a prior p(s) is used via the Bayes rule
``` p( x | s ) = p( s | x ) / p( s )  *  p( x ) ```
to replace the GMM emission prob. by a neural network.

Not sure if **only** needed here, but restricting it to the Gaussian Mixture Model (GMM) System is imho wrong, because it is also used in the Neural Network (NN) System.

>Is this needed for CTC/Transducer or some random new approaches? 

Transducer is a discriminative approach, so in principle no prior is needed. CTC can be both, a special case of a hybrid system or a direct discriminative model. Factored-Hybrid will definitively use it. Also currently we are supposed to do some Experiments that include priors into Transducers - but here I would strongly suggest to not use single-density-mixture but the softmax prior.

Reply:
Ok I see, then it can stay here. Thanks for the explanation!